types/mongoose/index.d.ts

```diff
@@ -983,13 +899,26 @@ declare module "mongoose" {
     }
 
     interface ValidateFn<T> {
-      (obj: RegExp | Function, message?: string, type?: string): T;
+      (value: T): boolean;
+    }
+
+    interface AsyncValidateFn<T> {
+      (value: T, done: (result: boolean) => void): void;
+    }
+
+    interface ValidateOptsBase {
+      msg?: string;
+      type?: string;
     }
 
-    interface ValidateOpts {
-      validator?: RegExp | Function,
-      msg?: string,
-      type?: string
+    interface ValidateOpts extends ValidateOptsBase {
+      isAsync?: false;
```

Comment:
Should this be `isAsync?: boolean`?

Reply:
This is to differentiate between `ValidatetOpts` and `AsyncValidateOpts`.

For ValidateOpts you need `isAsync` to be false or omitted and to provide a synchronous validation function.
For async validation you *need* to set `isAsync` to true and provide an async validation function. Marking the parameter as optional/non-optional and explicitly true/false instead of just boolean clearly expresses this logic.

Reply:
Got it, good idea.